src/coreclr/jit/assertionprop.cpp

```diff
@@ -4619,6 +4619,29 @@ GenTree* Compiler::optAssertionPropGlobal_RelOp(ASSERT_VALARG_TP assertions,
         return optAssertionProp_Update(newTree, tree, stmt);
     }
 
+    // See if we can fold "X relop CNS" using RangeCheck.
+    if (tree->OperIsCmpCompare() &&
+        // We can also use a more powerful SSA-based TryGetRange, but it's too expensive to call for every relop.
+        // Hence, the following checks are driven by TP/CQ balance:
+        op1->TypeIs(TYP_INT) && op2->IsIntCnsFitsInI32() && !op2->IsIntegralConst(0))
```

Comment:
Could this use op2's VN instead relying on it being a literal constant?

Reply:
Sure, I just assumed that we always propagate constants by value before global assertprop so while we generally should rely on VNs, for constants it's more or less fine 🙂 but in the sake of consistency I guess I can

Reply:
@AndyAyersMS I've addressed your feedback in https://github.com/dotnet/runtime/pull/112824/commits/877112cd69840dc9b9da1276115fe8c50dc3a1e9 and update the issue description